x-pack/legacy/plugins/infra/public/components/logging/log_text_stream/loading_item_view.tsx

```diff
@@ -28,122 +44,202 @@ export class LogTextStreamLoadingItemView extends React.PureComponent<
-                id="xpack.infra.logs.loadAgainButtonLabel"
-                defaultMessage="Load again"
-              />
-            </EuiButtonEmpty>
-          ) : null}
-        </ProgressEntry>
-      );
-    } else {
-      return null;
-    }
+    const shouldShowCta = !hasMore && !isStreaming;
+
+    const extra = (
+      <EuiFlexGroup justifyContent="center" alignItems="center" gutterSize="m">
+        {isLoading || isStreaming ? (
+          <ProgressSpinner kind={isStreaming ? 'streaming' : 'loading'} />
+        ) : shouldShowCta ? (
+          <ProgressCta
+            position={position}
+            onStreamStart={onStreamStart}
+            onExtendRange={onExtendRange}
+            rangeEdge={rangeEdge}
+          />
+        ) : null}
+      </EuiFlexGroup>
+    );
+
+    return (
+      <ProgressEntryWrapper className={className} position={position}>
+        {position === 'start' ? <>{extra}</> : null}
+        <ProgressMessage timestamp={timestamp} />
+        {position === 'end' ? <>{extra}</> : null}
+      </ProgressEntryWrapper>
+    );
   }
 }
 
-interface ProgressEntryProps {
-  alignment: 'top' | 'bottom';
-  className?: string;
-  color: 'subdued' | 'primary';
-  isLoading: boolean;
-}
+const ProgressEntryWrapper = euiStyled.div<{ position: Position }>`
+  padding-left: ${props => props.theme.eui.euiSizeS};
+  padding-top: ${props =>
+    props.position === 'start' ? props.theme.eui.euiSizeL : props.theme.eui.euiSizeM};
+  padding-bottom: ${props =>
+    props.position === 'end' ? props.theme.eui.euiSizeL : props.theme.eui.euiSizeM};
+`;
+
+const ProgressMessage: React.FC<{ timestamp?: number }> = ({ timestamp }) => {
+  return (
+    <LogTextSeparator>
+      <EuiTitle size="xxs">
+        {timestamp ? (
+          <FormattedMessage
+            id="xpack.infra.logs.showingEntriesUntilTimestamp"
+            defaultMessage="Showing entries until {timestamp}"
+            values={{ timestamp: <FormattedTime value={timestamp} /> }}
+          />
+        ) : (
+          <FormattedMessage
+            id="xpack.infra.logs.noAdditionalEntriesFoundText"
+            defaultMessage="No additional entries found"
+          />
+        )}
+      </EuiTitle>
+    </LogTextSeparator>
+  );
+};
+
+const ProgressSpinner: React.FC<{ kind: 'streaming' | 'loading' }> = ({ kind }) => (
+  <>
+    <EuiFlexItem grow={false}>
+      <EuiLoadingSpinner size="l" />
+    </EuiFlexItem>
+    <EuiFlexItem grow={false}>
+      <EuiText size="s">
+        {kind === 'streaming' ? (
+          <FormattedMessage
+            id="xpack.infra.logs.streamingNewEntriesText"
+            defaultMessage="Streaming new entries"
+          />
+        ) : (
+          <FormattedMessage
+            id="xpack.infra.logs.loadingNewEntriesText"
+            defaultMessage="Loading new entries"
+          />
+        )}
+      </EuiText>
+    </EuiFlexItem>
+  </>
+);
+
+const ProgressCta: React.FC<Pick<
+  LogTextStreamLoadingItemViewProps,
```

Comment:
I'm thinking out loud.

I understand the problem is that the type signature in the function is very complex. I agree with you that it's not easy to read what's going on.

I want to extract this into a type/interface as you requested, but I wonder what's the best way to do it.

#### As a standalone interface.

This is the existing proposal. We add the props to `ProgressCtaProps` and we compose `LogTextStreamLoadingItemViewProps` with it

```tsx
type Position = 'start' | 'end';

interface ProgressCtaProps {
  position: Position;
  /** startDate || endDate */
  rangeEdge?: string;
  onExtendRange?: (newDate: string) => void;
  onStreamStart?: () => void;
}

interface LogTextStreamLoadingItemViewProps extends ProgressCtaProps {
  /** topCursor.time || bottomCursor.time */
  timestamp?: number;
  className?: string;
  hasMore: boolean;
  isLoading: boolean;
  isStreaming: boolean;
}
```

#### As a selection of properties from the _main_ interface

This is similar to what exists in code right now, but extracting the complex part to a new type.

```tsx
type Position = 'start' | 'end';

interface LogTextStreamLoadingItemViewProps {
  position: Position;
  /** topCursor.time || bottomCursor.time */
  timestamp?: number;
  /** startDate || endDate */
  rangeEdge?: string;
  className?: string;
  hasMore: boolean;
  isLoading: boolean;
  isStreaming: boolean;
  onExtendRange?: (newDate: string) => void;
  onStreamStart?: () => void;
}

type ProgressCtaProps = Pick<LogTextStreamLoadingItemViewProps, 'position' | 'rangeEdge' | 'onExtendRange' | 'onStreamStart'>
```

---- 

I'm personally more inclined to the second option. The public component in the file is `LogTextStreamLoadingItemView`, and because of that I think its props should be the easier to read. If we split the props for that component in two interfaces I feel that makes it slightly harder to see how to use the component, since that information is now spread in two places.

The `ProgressCta` component is an implementation detail and is not exposed. It takes a subset of the props from the parent component. This relationship is reflected with the use of `Pick<>`.

A third possible option could be to just duplicate the props. That's [better than the wrong abstraction](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction).

Makes sense? What do you think? If you think it's more clear the other way I'll change it.